### ISSUE TITLE
Support for AWS SAM

### ISSUE THREAD

**Issue by ryanwilliams83** [jakejscott/Humidifier#7]
How would I go about adding support for AWS SAM into your library?

Basically it's a kind of self-expanding CloudFormation template transform that introduces new CLoudFormation types such as AWS::Serverless::Api and AWS::Serverless::Function

https://docs.aws.amazon.com/lambda/latest/dg/serverless_app.html

I see that you're using the namespace to form the Type attribute by replacing Humidifier with "AWS::" so I'm not sure how I can inherit from your classes to add the missing functionality.

Ultimately I'm attempting to insert an AWS::Serverless::Api entry with a "DefinitionBody" property set to a json swagger definition string

**Comment by jakejscott** [comment]
Hi @ryanwilliams83 thanks for the suggestion! I think it should be pretty easy to support.

The stack class does have the 'transform' property on it that you need to use to say that this template should be transformed: `AWSTemplateFormatVersion: '2010-09-09'`

https://github.com/jakejscott/Humidifier/blob/master/src/Humidifier/Stack.cs#L21

Then I think it's a matter of implementing the resource types from the spec here:
https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md

Also the the event source types:
https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#event-source-types

I think the types should go in a directory called `/src/Humidifier/Serverless`

If you give me a couple days I can implement it for you.

**Comment by jakejscott** [comment]
I also want to provide some sample projects, so I will add one for SAM first.

I'm also planning on creating a dotnet cli tooling similar to this: 
https://github.com/aws/aws-extensions-for-dotnet-cli#aws-lambda-amazonlambdatools

**Comment by ryanwilliams83** [comment]
I've modified my fork to have an AWSTypeName getter on all resources and modified the Json serializer to use it instead of derriving it from the namespace. Also modified the json serializer to omit this new property.

https://github.com/Busivid/Humidifier/commit/90d3a296aece75c67840be765ed78aa38307e5c2

**Comment by ryanwilliams83** [comment]
To make the framework extensible I added the AWSTypeName property to the Resource class
I was then able to inherit from Humudifier.Resource in my own namespace and create my own representations of the Serverless objects.

This pull request includes the changes to the Generator, the JSON Serializer and the Resource class only.
It fails the automated tests because it doesn't include all of the generated code files.
https://github.com/jakejscott/Humidifier/pull/8

**Comment by jakejscott** [comment]
Do you want to make all the classes implement the AWSTypeName so that it builds?

I think I'm happy with your approach. I've been wondering about how to make it extensible so nice work!

**Comment by jakejscott** [comment]
@ryanwilliams83 I've just pushed a new 0.2.5 version of the library.

I also added a stub for implementing the rest of the serverlss resource types: https://github.com/jakejscott/Humidifier/blob/master/src/Humidifier/Serverless/Function.cs

Thanks for the contribution :)

**Comment by WedioDK** [comment]
@jakejscott Would it be possible for you to fully support the SAM model as defined by AWS? E.g. all types.

https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-template.html#serverless-sam-template-application

**Comment by jakejscott** [comment]
What's missing? I added some more properties to this class a while back but just manually....

I found out that there is some JSON that describes the serverless cloudformation specs here:
https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/cfnspec/spec-source/000_sam.spec.json

But I just haven't had time to implement it as I don't use the Serverless transforms stuff and prefer to use the underlying resources directly (as Serverless templates just get expanded out in Cloudformation).

Feel free to let me know what's missing, or a PR and I can take a look.

Cheers
